File: library/src/main/java/com/xwray/groupie/Section.java
package com.xwray.groupie;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListUpdateCallback;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * A group which has a list of contents and an optional header and footer.
 */
public class Section extends NestedGroup {
    @Nullable
    private Group header;

    @Nullable
    private Group footer;

    @Nullable
    private Group placeholder;

    private final ArrayList<Group> children = new ArrayList<>();

    private boolean hideWhenEmpty = false;

    private boolean isHeaderAndFooterVisible = true;

    private boolean isPlaceholderVisible = false;

    public Section() {
        this(null, new ArrayList<Group>());
    }

    public Section(@Nullable Group header) {
        this(header, new ArrayList<Group>());
    }

    public Section(@NonNull Collection<? extends Group> children) {
        this(null, children);
    }

    public Section(@Nullable Group header, @NonNull Collection<? extends Group> children) {
        this.header = header;
        if (header != null) {
            header.registerGroupDataObserver(this);
        }
        addAll(children);
    }

    @Override
    public void add(int position, @NonNull Group group) {
        super.add(position, group);
        children.add(position, group);
        final int notifyPosition = getHeaderItemCount() + GroupUtils.getItemCount(children.subList(0, position));
        notifyItemRangeInserted(notifyPosition, group.getItemCount());
        refreshEmptyState();
    }

    @Override
    public void addAll(@NonNull Collection<? extends Group> groups) {
        if (groups.isEmpty()) return;
        super.addAll(groups);
        int position = getItemCountWithoutFooter();
        this.children.addAll(groups);
        notifyItemRangeInserted(position, GroupUtils.getItemCount(groups));
        refreshEmptyState();
    }

    @Override
    public void addAll(int position, @NonNull Collection<? extends Group> groups) {
        if (groups.isEmpty()) {
            return;
        }

        super.addAll(position, groups);
        this.children.addAll(position, groups);

        final int notifyPosition = getHeaderItemCount() + GroupUtils.getItemCount(children.subList(0, position));
        notifyItemRangeInserted(notifyPosition, GroupUtils.getItemCount(groups));
        refreshEmptyState();
    }

    @Override
    public void add(@NonNull Group group) {
        super.add(group);
        int position = getItemCountWithoutFooter();
        children.add(group);
        notifyItemRangeInserted(position, group.getItemCount());
        refreshEmptyState();
    }

    @Override
    public void remove(@NonNull Group group) {
        super.remove(group);
        int position = getItemCountBeforeGroup(group);
        children.remove(group);
        notifyItemRangeRemoved(position, group.getItemCount());
        refreshEmptyState();
    }

    @Override
    public void removeAll(@NonNull Collection<? extends Group> groups) {
        if (groups.isEmpty()) {
            return;
        }

        super.removeAll(groups);
        for (Group group : groups) {
            int position = getItemCountBeforeGroup(group);
            children.remove(group);
            notifyItemRangeRemoved(position, group.getItemCount());
        }
        refreshEmptyState();
    }

    @Override
    public void replaceAll(@NonNull Collection<? extends Group> groups) {
        if (groups.isEmpty()) {
            return;
        }

        super.replaceAll(groups);
        children.clear();
        children.addAll(groups);

        notifyDataSetInvalidated();

        refreshEmptyState();
    }

    /**
     * Get the list of all groups in this section, wrapped in a new {@link ArrayList}. This
     * does <strong>not include headers, footers or placeholders</strong>.
     * @return The list of all groups in this section, wrapped in a new {@link ArrayList}
     */
    public List<Group> getGroups() {
        return new ArrayList<>(this.children);
    }

    /**
     * Remove all existing body content.
     */
    public void clear() {
        if (children.isEmpty()) {
            return;
        }
        removeAll(new ArrayList<>(children));
    }

    /**
     * Replace all existing body content and dispatch fine-grained change notifications to the
     * parent using DiffUtil.
     * <p>
     * Item comparisons are made using:
     * - Item.isSameAs(Item otherItem) (are items the same?)
     * - Item.hasSameContentAs() (are contents the same?)
     * <p>
     * If you don't customize getId() or isSameAs() and hasSameContentAs(), the default implementations will return false,
     * meaning your Group will consider every update a complete change of everything.
     *
     * This will default detectMoves to true.
     *
     * @see #update(Collection, boolean)
     * @param newBodyGroups The new content of the section
     */
    public void update(@NonNull final Collection<? extends Group> newBodyGroups) {
        update(newBodyGroups, true);
    }

    /**
     * Replace all existing body content and dispatch fine-grained change notifications to the
     * parent using DiffUtil.
     * <p>
     * Item comparisons are made using:
     * - Item.isSameAs(Item otherItem) (are items the same?)
     * - Item.hasSameContentAs() (are contents the same?)
     * <p>
     * If you don't customize getId() or isSameAs() and hasSameContentAs(), the default implementations will return false,
     * meaning your Group will consider every update a complete change of everything.
     *
     * @param newBodyGroups The new content of the section
     * @param detectMoves is passed to {@link DiffUtil#calculateDiff(DiffUtil.Callback, boolean)}. Set to false if you
     *                    don't want DiffUtil to detect moved items.
     */
    public void update(@NonNull final Collection<? extends Group> newBodyGroups, boolean detectMoves) {
        final List<Group> oldBodyGroups = new ArrayList<>(children);
        final DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(new DiffCallback(oldBodyGroups, newBodyGroups), detectMoves);
        this.update(newBodyGroups, diffResult);
    }

    /**
     * Overloaded version of update method in which you can pass your own DiffUtil.DiffResult
     * @param newBodyGroups The new content of the section
     * @param diffResult
     */
    public void update(@NonNull final Collection<? extends Group> newBodyGroups, DiffUtil.DiffResult diffResult) {
        super.removeAll(children);
        children.clear();
        children.addAll(newBodyGroups);
        super.addAll(newBodyGroups);

        diffResult.dispatchUpdatesTo(listUpdateCallback);
        refreshEmptyState();
    }

    private ListUpdateCallback listUpdateCallback = new ListUpdateCallback() {
        @Override
        public void onInserted(int position, int count) {
            notifyItemRangeInserted(getHeaderItemCount() + position, count);
        }

        @Override
        public void onRemoved(int position, int count) {
            notifyItemRangeRemoved(getHeaderItemCount() + position, count);
        }

        @Override
        public void onMoved(int fromPosition, int toPosition) {
            final int headerItemCount = getHeaderItemCount();
            notifyItemMoved(headerItemCount + fromPosition, headerItemCount + toPosition);
        }

        @Override
        public void onChanged(int position, int count, Object payload) {
            notifyItemRangeChanged(getHeaderItemCount() + position, count, payload);
        }
    };

    /**
     * Optional. Set a placeholder for when the section's body is empty.
     * <p>
     * If setHideWhenEmpty(true) is set, then the empty placeholder will not be shown.
     *
     * @param placeholder A placeholder to be shown when there is no body content
     */
    public void setPlaceholder(@NonNull Group placeholder) {
        //noinspection ConstantConditions
        if (placeholder == null)
            throw new NullPointerException("Placeholder can't be null.  Please use removePlaceholder() instead!");
        if (this.placeholder != null) {
            removePlaceholder();
        }
        this.placeholder = placeholder;
        refreshEmptyState();
    }

    public void removePlaceholder() {
        hidePlaceholder();
        this.placeholder = null;
    }

    private void showPlaceholder() {
        if (isPlaceholderVisible || placeholder == null) return;
        isPlaceholderVisible = true;
        notifyItemRangeInserted(getHeaderItemCount(), placeholder.getItemCount());
    }

    private void hidePlaceholder() {
        if (!isPlaceholderVisible || placeholder == null) return;
        isPlaceholderVisible = false;
        notifyItemRangeRemoved(getHeaderItemCount(), placeholder.getItemCount());
    }

    /**
     * Whether a section's contents are visually empty
     *
     * @return
     */
    protected boolean isEmpty() {
        return children.isEmpty() || GroupUtils.getItemCount(children) == 0;
    }

    private void hideDecorations() {
        if (!isHeaderAndFooterVisible && !isPlaceholderVisible) return;
        int count = getHeaderItemCount() + getPlaceholderItemCount() + getFooterItemCount();
        isHeaderAndFooterVisible = false;
        isPlaceholderVisible = false;
        notifyItemRangeRemoved(0, count);
    }

    protected void refreshEmptyState() {
        boolean isEmpty = isEmpty();
        if (isEmpty) {
            if (hideWhenEmpty) {
                hideDecorations();
            } else {
                showPlaceholder();
                showHeadersAndFooters();
            }
        } else {
            hidePlaceholder();
            showHeadersAndFooters();
        }
    }

    private void showHeadersAndFooters() {
        if (isHeaderAndFooterVisible) return;
        isHeaderAndFooterVisible = true;
        notifyItemRangeInserted(0, getHeaderItemCount());
        notifyItemRangeInserted(getItemCountWithoutFooter(), getFooterItemCount());
    }

    private int getBodyItemCount() {
        return isPlaceholderVisible ? getPlaceholderItemCount() : GroupUtils.getItemCount(children);
    }

    private int getItemCountWithoutFooter() {
        return getBodyItemCount() + getHeaderItemCount();
    }

    private int getHeaderCount() {
        return header == null || !isHeaderAndFooterVisible ? 0 : 1;
    }

    private int getHeaderItemCount() {
        return getHeaderCount() == 0 ? 0 : header.getItemCount();
    }

    private int getFooterItemCount() {
        return getFooterCount() == 0 ? 0 : footer.getItemCount();
    }

    private int getFooterCount() {
        return footer == null || !isHeaderAndFooterVisible ? 0 : 1;
    }

    private int getPlaceholderCount() {
        return isPlaceholderVisible ? 1 : 0;
    }

    @Override
    @NonNull
    public Group getGroup(int position) {
        if (isHeaderShown() && position == 0) return header;
        position -= getHeaderCount();
        if (isPlaceholderShown() && position == 0) return placeholder;
        position -= getPlaceholderCount();
        if (position == children.size()) {
            if (isFooterShown()) {
                return footer;
            } else {
                throw new IndexOutOfBoundsException("Wanted group at position " + position +
                        " but there are only " + getGroupCount() + " groups");
            }
        } else {
            return children.get(position);
        }
    }

    @Override
    public int getGroupCount() {
        return getHeaderCount() + getFooterCount() + getPlaceholderCount() + children.size();
    }

    @Override
    public int getPosition(@NonNull Group group) {
        int count = 0;
        if (isHeaderShown()) {
            if (group == header) return count;
        }
        count += getHeaderCount();
        if (isPlaceholderShown()) {
            if (group == placeholder) return count;
        }
        count += getPlaceholderCount();

        int index = children.indexOf(group);
        if (index >= 0) return count + index;
        count += children.size();

        if (isFooterShown()) {
            if (footer == group) {
                return count;
            }
        }

        return -1;
    }

    private boolean isHeaderShown() {
        return getHeaderCount() > 0;
    }

    private boolean isFooterShown() {
        return getFooterCount() > 0;
    }

    private boolean isPlaceholderShown() {
        return getPlaceholderCount() > 0;
    }

    public void setHeader(@NonNull Group header) {
        if (header == null)
            throw new NullPointerException("Header can't be null.  Please use removeHeader() instead!");
        if (this.header != null) {
            this.header.unregisterGroupDataObserver(this);
        }
        int previousHeaderItemCount = getHeaderItemCount();
        this.header = header;
        header.registerGroupDataObserver(this);
        notifyHeaderItemsChanged(previousHeaderItemCount);
    }

    public void removeHeader() {
        if (this.header == null) {
            return;
        }

        this.header.unregisterGroupDataObserver(this);
        int previousHeaderItemCount = getHeaderItemCount();
        this.header = null;
        notifyHeaderItemsChanged(previousHeaderItemCount);
    }

    private void notifyHeaderItemsChanged(int previousHeaderItemCount) {
        int newHeaderItemCount = getHeaderItemCount();
        if (previousHeaderItemCount > 0) {
            notifyItemRangeRemoved(0, previousHeaderItemCount);
        }
        if (newHeaderItemCount > 0) {
            notifyItemRangeInserted(0, newHeaderItemCount);
        }
    }


    public void setFooter(@NonNull Group footer) {
        if (footer == null)
            throw new NullPointerException("Footer can't be null.  Please use removeFooter() instead!");
        if (this.footer != null) {
            this.footer.unregisterGroupDataObserver(this);
        }
        int previousFooterItemCount = getFooterItemCount();
        this.footer = footer;
        footer.registerGroupDataObserver(this);
        notifyFooterItemsChanged(previousFooterItemCount);
    }

    public void removeFooter() {
        if (this.footer == null) {
            return;
        }

        this.footer.unregisterGroupDataObserver(this);
        int previousFooterItemCount = getFooterItemCount();
        this.footer = null;
        notifyFooterItemsChanged(previousFooterItemCount);
    }

    private void notifyFooterItemsChanged(int previousFooterItemCount) {
        int newFooterItemCount = getFooterItemCount();
        if (previousFooterItemCount > 0) {
            notifyItemRangeRemoved(getItemCountWithoutFooter(), previousFooterItemCount);
        }
        if (newFooterItemCount > 0) {
            notifyItemRangeInserted(getItemCountWithoutFooter(), newFooterItemCount);
        }
    }

    public void setHideWhenEmpty(boolean hide) {
        if (hideWhenEmpty == hide) return;
        hideWhenEmpty = hide;
        refreshEmptyState();
    }

    @Override
    public void onItemInserted(@NonNull Group group, int position) {
        super.onItemInserted(group, position);
        refreshEmptyState();
    }

    @Override
    public void onItemRemoved(@NonNull Group group, int position) {
        super.onItemRemoved(group, position);
        refreshEmptyState();
    }

    @Override
    public void onItemRangeInserted(@NonNull Group group, int positionStart, int itemCount) {
        super.onItemRangeInserted(group, positionStart, itemCount);
        refreshEmptyState();
    }

    @Override
    public void onItemRangeRemoved(@NonNull Group group, int positionStart, int itemCount) {
        super.onItemRangeRemoved(group, positionStart, itemCount);
        refreshEmptyState();
    }

    private int getPlaceholderItemCount() {
        if (isPlaceholderVisible && placeholder != null) {
            return placeholder.getItemCount();
        }
        return 0;
    }
}

File: library-kotlin/src/main/java/com/wray/groupiekotlin/Section.kt
package com.wray.groupiekotlin

import androidx.annotation.CallSuper
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListUpdateCallback

@Suppress("LeakingThis")
open class Section(
    private var header: Group? = null,
    private var footer: Group? = null,
    private var placeholder: Group? = null,
    children: List<Group> = emptyList()
) : NestedGroup() {

    private val children = mutableListOf<Group>()
    private var isPlaceholderVisible = false
    private var isHeaderAndFooterVisible = true

    /**
     * Whether this entire section should be hidden when empty. If false, it will
     * still show any footers, headers or placeholders set if empty.
     */
    var hideWhenEmpty = false
        set(value) {
            if (field == value) return
            field = value
            refreshEmptyState()
        }

    /**
     * Whether this section's contents are visually empty.
     */
    protected val isEmpty: Boolean
        get() = children.isEmpty() || children.itemCount == 0

    private val headerCount: Int
        get() = if (header == null || !isHeaderAndFooterVisible) 0 else 1
    private val headerItemCount: Int
        get() = if (headerCount == 0) 0 else header?.itemCount ?: 0
    private val isHeaderShown: Boolean
        get() = headerCount > 0

    private val footerCount: Int
        get() = if (footer == null || !isHeaderAndFooterVisible) 0 else 1
    private val footerItemCount: Int
        get() = if (footerCount == 0) 0 else footer?.itemCount ?: 0
    private val isFooterShown: Boolean
        get() = footerCount > 0

    private val placeholderCount: Int
        get() = if (isPlaceholderVisible) 1 else 0
    private val placeholderItemCount: Int
        get() = if (!isPlaceholderVisible || placeholder == null) 0 else placeholder?.itemCount ?: 0
    private val isPlaceholderShown: Boolean
        get() = placeholderCount > 0

    private val bodyItemCount: Int
        get() = if (isPlaceholderVisible) placeholderItemCount else children.itemCount

    private val itemCountWithoutFooter: Int
        get() = bodyItemCount + headerItemCount

    private val listUpdateCallback = object : ListUpdateCallback {
        override fun onInserted(position: Int, count: Int) {
            notifyItemRangeInserted(headerItemCount + position, count)
        }

        override fun onRemoved(position: Int, count: Int) {
            notifyItemRangeRemoved(headerItemCount + position, count)
        }

        override fun onMoved(fromPosition: Int, toPosition: Int) {
            val headerItemCount = headerItemCount
            notifyItemMoved(headerItemCount + fromPosition, headerItemCount + toPosition)
        }

        override fun onChanged(position: Int, count: Int, payload: Any?) {
            if (payload != null) {
                notifyItemRangeChanged(headerItemCount + position, count, payload)
            } else {
                notifyItemRangeChanged(headerItemCount + position, count)
            }
        }
    };

    init {
        header?.registerGroupDataObserver(this)
        footer?.registerGroupDataObserver(this)
        placeholder?.registerGroupDataObserver(this)

        addAll(children)
    }

    override fun add(group: Group) {
        super.add(group)
        val position = itemCountWithoutFooter
        children.add(group)
        notifyItemRangeInserted(position, group.itemCount)
        refreshEmptyState()
    }

    override fun add(position: Int, group: Group) {
        super.add(position, group)
        children.add(position, group)
        val notifyPosition = headerItemCount + children.subList(0, position).itemCount
        notifyItemRangeInserted(notifyPosition, group.itemCount)
        refreshEmptyState()
    }

    override fun <T : Group> addAll(groups: Collection<T>) {
        if (groups.isEmpty()) return
        super.addAll(groups)

        val position = itemCountWithoutFooter
        children.addAll(groups)
        notifyItemRangeInserted(position, groups.itemCount)
        refreshEmptyState()
    }

    override fun <T : Group> addAll(position: Int, groups: Collection<T>) {
        if (groups.isEmpty()) return
        super.addAll(position, groups)
        children.addAll(position, groups)

        val notifyPosition = headerItemCount + children.subList(0, position).itemCount
        notifyItemRangeInserted(notifyPosition, groups.itemCount)
        refreshEmptyState()
    }

    override fun remove(group: Group) {
        super.remove(group)
        val position = getItemCountBeforeGroup(group)
        children.remove(group)
        notifyItemRangeRemoved(position, group.itemCount)
        refreshEmptyState()
    }

    override fun <T : Group> removeAll(groups: Collection<T>) {
        if (groups.isEmpty()) return
        super.removeAll(groups)

        groups.forEach { group ->
            val position = getItemCountBeforeGroup(group)
            children.remove(group)
            notifyItemRangeRemoved(position, group.itemCount)
        }

        refreshEmptyState()
    }

    /**
     * Replace all existing body content and dispatch fine-grained change notifications to the
     * parent using DiffUtil.
     *
     * Item comparisons are made using:
     * - `Item.isSameAs(Item otherItem)` (are items the same?)
     * - `Item.equals()` (are contents the same?)
     *
     * If you don't customize getId() or isSameAs() and equals(), the default implementations will return false,
     * meaning your Group will consider every update a complete change of everything.
     *
     * @param newBodyGroups The new content of the section
     * @param detectMoves is passed to [DiffUtil.calculateDiff(DiffUtil.Callback, boolean)]. Set to false if you
     *                    don't want DiffUtil to detect moved items.
     */
    fun update(newBodyGroups: List<Group>, detectMoves: Boolean = true) {
        val oldBodyGroups = children.toList()
        val oldBodyItemCount = oldBodyGroups.itemCount
        val newBodyItemCount = newBodyGroups.itemCount

        val diffResult = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
            override fun getOldListSize(): Int = oldBodyItemCount

            override fun getNewListSize(): Int = newBodyItemCount

            override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                val oldItem = oldBodyGroups.getItem(oldItemPosition)
                val newItem = newBodyGroups.getItem(newItemPosition)
                return newItem.isSameAs(oldItem)
            }

            override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                val oldItem = oldBodyGroups.getItem(oldItemPosition)
                val newItem = newBodyGroups.getItem(newItemPosition)
                return newItem == oldItem
            }
        }, detectMoves)

        super.removeAll(children)
        children.clear()
        children.addAll(newBodyGroups)
        super.addAll(newBodyGroups)

        diffResult.dispatchUpdatesTo(listUpdateCallback)
        if (newBodyItemCount == 0 || oldBodyItemCount == 0) {
            refreshEmptyState()
        }
    }

    /**
     * Overloaded version of update method in which you can pass your own [DiffUtil.DiffResult]
     */
    fun update(newBodyGroups: List<Group>, diffResult: DiffUtil.DiffResult) {
        super.removeAll(children)
        children.clear()
        children.addAll(newBodyGroups)
        super.addAll(newBodyGroups)

        diffResult.dispatchUpdatesTo(listUpdateCallback)
        refreshEmptyState()
    }

    /**
     * Remove all existing body content. Does not remove headers, footers or placeholders
     */
    @CallSuper
    open fun clear() {
        if (children.isEmpty()) {
            return
        }

        removeAll(children.toList())
    }

    fun setPlaceholder(placeholder: Group) {
        removePlaceholder()
        this.placeholder = placeholder
        refreshEmptyState()
    }

    fun removePlaceholder() {
        hidePlaceholder()
        placeholder = null
    }

    fun setHeader(header: Group) {
        val previousHeaderItemCount = headerItemCount
        this.header?.unregisterGroupDataObserver(this)
        this.header = header
        header.registerGroupDataObserver(this)
        notifyHeaderItemsChanged(previousHeaderItemCount)
    }

    fun removeHeader() {
        header?.let { header ->
            header.unregisterGroupDataObserver(this)
            val previousHeaderItemCount = headerItemCount
            this.header = null
            notifyHeaderItemsChanged(previousHeaderItemCount)
        }
    }

    fun setFooter(footer: Group) {
        val previousFooterItemCount = footerItemCount
        this.footer?.unregisterGroupDataObserver(this)
        this.footer = footer
        footer.registerGroupDataObserver(this)
        notifyFooterItemsChanged(previousFooterItemCount)
    }

    fun removeFooter() {
        footer?.let { footer ->
            footer.unregisterGroupDataObserver(this)
            val previousFooterItemCount = footerItemCount
            this.footer = null
            notifyFooterItemsChanged(previousFooterItemCount)
        }
    }

    private fun notifyHeaderItemsChanged(previousHeaderItemCount: Int) {
        val newHeaderItemCount = headerItemCount
        if (previousHeaderItemCount > 0) {
            notifyItemRangeRemoved(0, previousHeaderItemCount)
        }
        if (newHeaderItemCount > 0) {
            notifyItemRangeInserted(0, newHeaderItemCount)
        }
    }

    private fun notifyFooterItemsChanged(previousFooterItemCount: Int) {
        val newFooterItemCount = footerItemCount
        if (previousFooterItemCount > 0) {
            notifyItemRangeRemoved(itemCountWithoutFooter, previousFooterItemCount)
        }
        if (newFooterItemCount > 0) {
            notifyItemRangeInserted(itemCountWithoutFooter, newFooterItemCount)
        }
    }

    private fun showPlaceholder() {
        if (isPlaceholderVisible || placeholder == null)
            return

        isPlaceholderVisible = true
        notifyItemRangeInserted(headerItemCount, placeholder!!.itemCount)
    }

    private fun hidePlaceholder() {
        if (!isPlaceholderVisible || placeholder == null)
            return

        isPlaceholderVisible = true
        notifyItemRangeRemoved(headerItemCount, placeholder!!.itemCount)
    }

    private fun hideDecorations() {
        if (!isHeaderAndFooterVisible && !isPlaceholderVisible)
            return

        val count = headerItemCount + placeholderItemCount + footerItemCount
        isHeaderAndFooterVisible = false
        isPlaceholderVisible = false
        notifyItemRangeRemoved(0, count)
    }

    protected fun refreshEmptyState() {
        val isEmpty = isEmpty
        if (isEmpty) {
            if (hideWhenEmpty) {
                hideDecorations()
            } else {
                showPlaceholder()
                showHeadersAndFooters()
            }
        } else {
            hidePlaceholder()
            showHeadersAndFooters()
        }
    }

    private fun showHeadersAndFooters() {
        if (isHeaderAndFooterVisible)
            return

        isHeaderAndFooterVisible = true
        notifyItemRangeInserted(0, headerItemCount)
        notifyItemRangeInserted(itemCountWithoutFooter, footerItemCount)
    }

    //region: NestedGroup methods

    override fun getGroup(position: Int): Group {
        var pos = position

        if (isHeaderShown && pos == 0) {
            return header!!
        }
        pos -= headerCount

        if (isPlaceholderShown && pos == 0) {
            return placeholder!!
        }
        pos -= placeholderCount

        return if (pos == children.size) {
            if (isFooterShown) {
                footer!!
            } else {
                throw IndexOutOfBoundsException("Wanted group at position $pos but there are only $groupCount groups")
            }
        } else {
            children[pos]
        }
    }

    override val groupCount: Int
        get() = headerCount + footerCount + placeholderCount + children.size

    override fun getPosition(group: Group): Int {
        var count = 0

        if (isHeaderShown) {
            if (group == header) {
                return count
            }
        }
        count += headerCount

        if (isPlaceholderShown) {
            if (group == placeholder) {
                return count
            }
        }
        count += placeholderCount

        val index = children.indexOf(group)
        if (index >= 0) {
            return count + index
        }
        count += children.size

        if (isFooterShown && footer == group) {
            return count
        }

        return -1
    }

    //endregion: NestedGroup methods

    override fun onItemInserted(group: Group, position: Int) {
        super.onItemInserted(group, position)
        refreshEmptyState()
    }

    override fun onItemRemoved(group: Group, position: Int) {
        super.onItemRemoved(group, position)
        refreshEmptyState()
    }

    override fun onItemRangeInserted(group: Group, positionStart: Int, itemCount: Int) {
        super.onItemRangeInserted(group, positionStart, itemCount)
        refreshEmptyState()
    }

    override fun onItemRangeRemoved(group: Group, positionStart: Int, itemCount: Int) {
        super.onItemRangeRemoved(group, positionStart, itemCount)
        refreshEmptyState()
    }
}

private fun <T : Group> Collection<T>.getItem(position: Int): AnyItem {
    var previousPosition = 0
    for (group in this) {
        val size = group.itemCount
        if (size + previousPosition > position) {
            return group.getItem(position - previousPosition)
        }
        previousPosition += size
    }

    throw IndexOutOfBoundsException("Wanted item at $position but there are only $previousPosition items")
}